src/main/java/at/hannibal2/skyhanni/features/mining/eventtracker/MiningEventDisplay.kt

```diff
@@ -0,0 +1,92 @@
+import at.hannibal2.skyhanni.utils.LocationUtils
+import at.hannibal2.skyhanni.utils.LorenzUtils
+import at.hannibal2.skyhanni.utils.LorenzUtils.isInIsland
+import at.hannibal2.skyhanni.utils.RenderUtils.renderStrings
+import at.hannibal2.skyhanni.utils.SimpleTimeMark.Companion.asTimeMark
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+
+object MiningEventDisplay {
+    private val config get() = SkyHanniMod.feature.mining.miningEvent
+    private var display = mutableListOf<String>()
+
+    private val dwarvenEvents = mutableListOf<RunningEvent>()
+    private val crystalEvents = mutableListOf<RunningEvent>()
+    private var lastDwarvenEvent: MiningEvent? = null
+    private var lastCrystalEvent: MiningEvent? = null
+
+    @SubscribeEvent
+    fun onTick(event: LorenzTickEvent) {
+        if (!event.repeatSeconds(1)) return
+        updateDisplay()
+    }
+
+    @SubscribeEvent
+    fun onRenderOverlay(event: GuiRenderEvent.GuiOverlayRenderEvent) {
+        if (!shouldDisplay()) return
+        config.position.renderStrings(display, posLabel = "Upcoming Events Display")
+    }
+
+    private fun updateDisplay() {
+        display.clear()
+        updateEvents(IslandType.DWARVEN_MINES, dwarvenEvents, lastDwarvenEvent)
+        updateEvents(IslandType.CRYSTAL_HOLLOWS, crystalEvents, lastCrystalEvent)
+    }
+
+    private fun updateEvents(islandType: IslandType, events: List<RunningEvent>, lastEvent: MiningEvent?) {
+        val shouldShow = when (config.showType) {
+            MiningEventConfig.ShowType.DWARVEN -> islandType == IslandType.DWARVEN_MINES
+            MiningEventConfig.ShowType.CRYSTAL -> islandType == IslandType.CRYSTAL_HOLLOWS
+            MiningEventConfig.ShowType.CURRENT -> islandType.isInIsland()
+            else -> true
+        }
+
+        events.firstOrNull()?.let { firstEvent ->
+            if (firstEvent.endsAt.asTimeMark().isInPast()) {
+                when (islandType) {
+                    IslandType.DWARVEN_MINES -> lastDwarvenEvent = firstEvent.event
+                    IslandType.CRYSTAL_HOLLOWS -> lastCrystalEvent = firstEvent.event
+                    else -> Unit
+                }
+            }
+        }
+
+        if (shouldShow) {
+            val upcomingEvents = formatUpcomingEvents(events, lastEvent)
+            display.add("§a${islandType.displayName}§8: $upcomingEvents")
+        }
+    }
+
+    private fun formatUpcomingEvents(events: List<RunningEvent>, lastEvent: MiningEvent?): String {
+        val upcoming = events.filter { !it.endsAt.asTimeMark().isInPast() }
+            .map { if (it.isDouble) "${it.event} §8-> ${it.event}" else it.event.toString() }
+            .toMutableList()
+
+        if (upcoming.isEmpty()) upcoming.add("§7???")
+        if (config.passedEvents && upcoming.size < 4) lastEvent?.let { upcoming.add(0, it.toPastString()) }
+        return upcoming.joinToString(" §8-> ")
+    }
+
+    fun updateData(eventData: MiningEventData) {
```

Comment:
Since this function is being called async, id say its more consistent to change the lists dwarven events and crystal events from val mutable to var immutable. Also maybe work with DelayedRun for more consistency